src/utils/io/progress_log.rs

```diff
@@ -0,0 +1,235 @@
+    into_iter: impl IntoIterator<IntoIter = Inner>,
+) -> WithProgressIter<Inner>
+where
+    Inner: Iterator,
+{
+    let inner = into_iter.into_iter();
+    let (lower_bound, upper_bound) = inner.size_hint();
+    let total_items = upper_bound.unwrap_or(lower_bound) as u64;
+    WithProgressIter {
+        inner,
+        progress: WithProgress::new(message, total_items),
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct WithProgressIter<Inner> {
+    inner: Inner,
+    progress: WithProgress,
+}
+
+impl<Inner> Iterator for WithProgressIter<Inner>
+where
+    Inner: Iterator,
+{
+    type Item = Inner::Item;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.inner.next() {
+            Some(item) => {
+                self.progress.inc(1);
+                Some(item)
+            }
+            None => {
+                // TODO handle fusing
+                println!("finished {} items", self.progress.completed_items);
+                None
+            }
+        }
+    }
+}
+
+#[allow(dead_code)]
+pub fn wrap_stream<S: futures_core::Stream>(message: &str, stream: S) -> WithProgressStream<S> {
+    let (lower_bound, upper_bound) = stream.size_hint();
+    let total_items = upper_bound.unwrap_or(lower_bound) as u64;
+    WithProgressStream {
+        stream,
+        progress: WithProgress::new(message, total_items),
+    }
+}
+
+pub fn wrap_async_read<R: tokio::io::AsyncRead + Unpin>(
+    message: &str,
+    read: R,
+    total_items: u64,
+) -> WithProgressStream<R> {
+    WithProgressStream {
+        stream: read,
+        progress: WithProgress::new(message, total_items),
+    }
+}
+
+/// Wraps an iterator to display its progress.
+#[derive(Debug)]
+pub struct WithProgressStream<S> {
+    pub(crate) stream: S,
+    progress: WithProgress,
+}
+
+impl<R: tokio::io::AsyncRead + Unpin> tokio::io::AsyncRead for WithProgressStream<R> {
+    fn poll_read(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut ReadBuf<'_>,
+    ) -> Poll<io::Result<()>> {
+        let prev_len = buf.filled().len() as u64;
+        if let Poll::Ready(e) = Pin::new(&mut self.stream).poll_read(cx, buf) {
+            self.progress.inc(buf.filled().len() as u64 - prev_len);
+            Poll::Ready(e)
+        } else {
+            Poll::Pending
+        }
+    }
+}
+
+impl<W: tokio::io::AsyncBufRead + Unpin + tokio::io::AsyncRead> tokio::io::AsyncBufRead
+    for WithProgressStream<W>
+{
+    fn poll_fill_buf(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<io::Result<&[u8]>> {
+        let this = self.get_mut();
+        let result = Pin::new(&mut this.stream).poll_fill_buf(cx);
+        if let Poll::Ready(Ok(buf)) = &result {
+            this.progress.inc(buf.len() as u64);
+        }
+        result
+    }
+
+    fn consume(mut self: Pin<&mut Self>, amt: usize) {
+        Pin::new(&mut self.stream).consume(amt);
+    }
+}
+
+impl<S: futures_core::Stream + Unpin> futures_core::Stream for WithProgressStream<S> {
```

Comment:
You want to implement `size_hint` here as well.

Reply:
I removed this type since it was dead code.